### DISCUSSION TITLE
How to run any .mpy file ?

### DISCUSSION THREAD

**Discussion by zcattacz** [micropython/micropython#11633] [ESP32] [answered]
With .py , open()+read()+exec() is enough, what about mpy, provided that the mpy code structure is unknown ?
[This](https://stackoverflow.com/questions/29897480/possible-to-execute-python-bytecode-from-a-script) SO page suggests `marshal` to unmarshal bytecode into code object. which doesn't seem to be applicable in mpy.

**Comment by Josverl** [comment]
So that is a module contained in a single .mpy file

For `foo.mpy` you can load and execute it using `import foo`, which will l execute any code at the 'base level' ( not indented) in that module. 
Note that as you are importing a module, this can/needs to be done only once.

Assuming you have a function `main` in that file, you then can call this using `foo.main()`

**Comment by cwalther** [comment, marked as answer]
`import` is the correct way, and you can do it multiple times by removing the module from `sys.modules` in between (`del sys.modules['foo']`). If you have the module name as a string variable (rather than a hardcoded string like `'foo'` in the preceding example), you can use `__import__(name)`.

**Comment by zcattacz** [comment]
Thank you, import does feels like the practical way on mpy. 
On CPython, file path prevent arbitrary script from being imported, I forgot mpy is much simplified.

**Comment by jimmo** [comment]
If you have the file only in RAM, then you can build a in-RAM filesystem to allow the import to find it. But it sounds like from your question that the file is already on the filesystem, so `import` is definitely the way to go.

You might also be interested in https://github.com/micropython/micropython/pull/8744  (but again, you should use `import` :) )

**Reply by zcattacz** [reply]
Thank you for the tip, exactly, in RAM was what I was looking for. Changed the plan for now, since I have no clue how :-).
